File: test/language/templates/cpp98/StructRecursiveTemplateTest.cpp
#include "gtest/gtest.h"

#include "templates/struct_recursive_template/StructRecursiveTemplate.h"

namespace templates
{
namespace struct_recursive_template
{

TEST(StructRecursiveTemplateTest, readWrite)
{
    StructRecursiveTemplate structRecursiveTemplate;
    structRecursiveTemplate.getCompound1().getValue().setValue(42);
    structRecursiveTemplate.getCompound2().getValue().getValue().setValue("string");

    zserio::BitStreamWriter writer;
    structRecursiveTemplate.write(writer);
    size_t bufferSize = 0;
    const uint8_t* buffer = writer.getWriteBuffer(bufferSize);

    zserio::BitStreamReader reader(buffer, bufferSize);
    StructRecursiveTemplate readStructRecursiveTemplate(reader);

    ASSERT_TRUE(structRecursiveTemplate == readStructRecursiveTemplate);
}

} // namespace struct_recursive_template
} // namespace templates

File: test/extensions/language/templates/java/templates/StructRecursiveTemplateTest.java
package templates;

import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;

import org.junit.jupiter.api.Test;

import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

import templates.struct_recursive_template.Compound_Compound_Compound_string;
import templates.struct_recursive_template.Compound_Compound_string;
import templates.struct_recursive_template.Compound_Compound_uint32;
import templates.struct_recursive_template.Compound_string;
import templates.struct_recursive_template.Compound_uint32;
import templates.struct_recursive_template.StructRecursiveTemplate;

public class StructRecursiveTemplateTest
{
    @Test
    public void readWrite() throws IOException
    {
        final StructRecursiveTemplate structRecursiveTemplate = new StructRecursiveTemplate();
        structRecursiveTemplate.setCompound1(new Compound_Compound_uint32(new Compound_uint32(42)));
        structRecursiveTemplate.setCompound2(new Compound_Compound_Compound_string(
                new Compound_Compound_string(new Compound_string("string"))));

        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        structRecursiveTemplate.write(writer);

        final BitStreamReader reader =
                new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        final StructRecursiveTemplate readStructRecursiveTemplate = new StructRecursiveTemplate(reader);
        assertTrue(structRecursiveTemplate.equals(readStructRecursiveTemplate));
    }
}
